client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,77 +20,83 @@ import {
-			auth_type: authType,
-			two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
-			remember_me: getRememberMe( store.getState() ),
-			two_step_push_token: getTwoFactorPushToken( store.getState() ),
-			client_id: config( 'wpcom_signup_id' ),
-			client_secret: config( 'wpcom_signup_key' ),
-		} ).then( () => {
-			store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
-		} ).catch( error => {
-			store.dispatch( {
-				type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
-				nonceType: 'push',
-				twoStepNonce: error.response.body.data.two_step_nonce
-			} );
-			return Promise.reject( error );
-		} );
+	store.dispatch(
+		http( {
+			url: 'https://wordpress.com/wp-login.php?action=two-step-authentication-endpoint',
+			method: 'POST',
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
+	}
+
+	store.dispatch( {
+		type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
+		nonceType: 'push',
+		twoStepNonce,
+	} );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
```

Comment:
> Why it needs to be in the handler? the UI should do whatever it wants when the operation succeeds,
> we shouldn't decide about it here.

that wasn't a strong statement I was making. the point was that this is the spot for side effects and maybe that's all that mattered. ideally we stop using `page()` in React components anyway and just dispatch an action requesting the navigation change but that's still far off in the future before we get there.

the signaling to the UI doesn't have to be "this polling operation has finished" - it can be "we have the authorized 2fa" which takes out the need for the UI to care what happens in response. but yeah, it could sit there and watch for the changed state and respond…

```js
componentShouldUpdate( prevProps ) {
	if ( ( ! prevProps.hasValid2FAuth ) && this.props.hasValid2FAuth ) {
		// respond
	}
}
```